potassium-nitrite/src/main/kotlin/org/dizitart/kno2/serialization/DocumentFormat.kt

```diff
@@ -0,0 +1,174 @@
+    fun <T : Any> decodeFromDocument(serializer: KSerializer<T>, document: Document): T =
+        json.decodeFromJsonElement(serializer, document.toJsonObject())
+
+    internal class Custom(override val json: Json) : DocumentFormat()
+
+}
+
+/**
+ * Type alias for the builder used to configure a [DocumentFormat] instance.
+ */
+typealias DocumentFormatBuilder = JsonBuilder
+
+/**
+ * Builder function to create a [DocumentFormat] instance with a custom configuration.
+ *
+ * @param from The base [DocumentFormat] to start with (defaults to [DocumentFormat.Default]).
+ * @param configure A lambda function to configure the [DocumentFormatBuilder].
+ * @return A new [DocumentFormat] instance with the specified configuration.
+ */
+fun DocumentFormat(
+    from: DocumentFormat = DocumentFormat.Default,
+    configure: DocumentFormatBuilder.() -> Unit,
+): DocumentFormat {
+    val jsonConfig = Json(from.json) {
+        configure()
+        ignoreUnknownKeys = true
+    }
+    return DocumentFormat.Custom(jsonConfig)
+}
+
+val DocumentFormat.serializersModule: SerializersModule
+    get() = json.serializersModule
+
+/**
+ * Encodes the given [value] of type [T] into a Nitrite [Document] using the serializer registered in the [serializersModule].
+ */
+inline fun <reified T : Any> DocumentFormat.encodeToDocument(value: T): Document =
+    encodeToDocument(json.serializersModule.serializer<T>(), value)
+
+/**
+ * Decodes the given Nitrite [document] into a value of type [T] using the serializer registered in the [serializersModule].
+ */
+inline fun <reified T> DocumentFormat.decodeFromDocument(document: Document): T =
+    decodeFromDocument(json.serializersModule.serializer(), document)
+
+private fun Document.toJsonObject() =
+    JsonObject(associate { (key, value) -> key to value.toJsonElement() })
+
+private fun Any?.toJsonElement(): JsonElement = when (this) {
+    is String -> JsonPrimitive(this)
+    is Number -> JsonPrimitive(this)
+    is Boolean -> JsonPrimitive(this)
+    is Document -> toJsonObject()
+    is List<*> -> JsonArray(map { it.toJsonElement() })
+    is Map<*, *> -> JsonObject(checkKeysAreString().mapValues { (_, value) -> value.toJsonElement() })
+    is Enum<*> -> JsonPrimitive(this.name)
+    null -> JsonNull
+    else -> throw ObjectMappingException("Can't convert object of type " + this::class.java + " to JSON")
+}
+
+private fun <K, V> Map<out K, V>.checkKeysAreString(): Map<String, V> = mapKeys { (key, _) ->
+    when (key) {
+        is String -> key
+        else -> throw ObjectMappingException("Map keys must be of type String")
+    }
+}
+
+private fun JsonElement.toDocument(): Any? = when (this) {
+    is JsonObject -> toDocument()
+    is JsonArray -> toDocument()
+    is JsonNull -> null
+    is JsonPrimitive -> toDocument()
+}
+
+private fun JsonObject.toDocument(): Document {
+    val document = Document.createDocument()
+    for ((key, value) in entries) {
+        document.put(key, value.toDocument(), true)
```

Comment:
This is very much your use case specific change. If some other user choose to use this kotlinx.serialization they won't be able to use/query nested documents. Please see my comments [here](https://github.com/nitrite/nitrite-java/pull/968#discussion_r1603752011).

Reply:
let's use [this comment](https://github.com/nitrite/nitrite-java/pull/968#discussion_r1603752011) to follow up the topic!